src/librustc/session/mod.rs

```diff
@@ -598,8 +598,12 @@ impl Session {
     }
 
     pub fn must_not_eliminate_frame_pointers(&self) -> bool {
-        self.opts.debuginfo != DebugInfoLevel::NoDebugInfo ||
-        !self.target.target.options.eliminate_frame_pointer
+        if let Some(x) = self.opts.cg.force_frame_pointers {
```

Comment:
To me, the name "force-frame-pointers" makes it sound like a "no" value would revert to default behavior -- but I guess that's not quite right. (That is, if you give `-Cforce-frame-pointer=no` and you have debuginfo, it will still strip your frame pointers, right? I think I would expect it not to based on the name.)

Reply:
`-Cforce-frame-pointer=no` to me sounds like we're asking the compiler to not force any frame pointers. This lets the compiler use frame pointers any way it pleases.

The other proposed names for this flag sounds like they are omitting or stripping frame pointers. LLVM does not support this. You can only ask it to force frame pointers or let it decide how frame pointers are used.

Reply:
> This lets the compiler use frame pointers any way it pleases.

I agree, it gives us freedom. I guess I'm saying is that what I would expect `-C force-frame-pointers=no` and not saying *anything* to be equivalent.

Reply:
To be clear, this is basically a bikeshed =) but it argues for @Mark-Simulacrum's "yes/no/auto" formulation.

Reply:
@nikomatsakis ah I see your point, however, it is something that could apply for all optionally-bool arguments, with `auto` being the same as if the argument was never specified anywhere in the CLI.

Reply:
As far as I understand it, you cannot completely opt out of frame pointer emission, you can only opt into always generating them, so a simple boolean flag `-C always-emit-frame-pointers` would suffice, I think. The name is a bit verbose though.

Reply:
I find @michaelwoerister's logic compelling, but I'm not sure it's complete. `C force-frame-pointers` is pretty clear -- if you provide it, it forces frame pointers, otherwise it falls back to our default behavior. I guess the question is whether we want the ability to allow frame-pointers to be optimized also in cases where we would by default include them (e.g., with debuginfo, or on mac, or whatever). 

I also find @Mark-Simulacrum's formulation compelling, but it sounds a bit like work. =) I guess we would say that `-Cframe-pointers=off` errors when you are not in a debug build?

OTOH, if we do go with `-Cforce-frame-pointers` as a simple boolean, we can clearly expand later and just keep it as a deprecated alias. It seems to commit us to very little. =)

Reply:
Okay, so…

I’d like to direct you all to read the commit message:

> This reworks the force-frame-pointer PR to explicitly only consider the
value of the flag if it is provided and also keeps the debug-info =
force-frame-pointer behaviour from before.
>
> Something that was tried but not kept was renaming the flag to
`frame-pointer`, because for flag `frame-pointer=no`, it is not
guaranteed, that no frame pointers will be generated. Rather, all it does
is inhibit the removal of the frame pointer.

This explains well why I didn’t choose `frame-pointers` wording.

`-C always emit-frame-pointers` seems like a synonym to `-C force-frame-pointers` to me.

I read `-C force-frame-pointers = no` as "Don’t force frame pointers”, rather than anything else, which aligns well with it simply allowing LLVM to omit the frame pointer anywhere it wants, but not guaranteeing that.

I also didn’t add `auto` at this moment, because it doesn’t seem to imply "default" to me, but I could also be persuaded otherwise. Anyway it is something that can be added later, and since it is more generally useful, considered for all the flags at once.

Reply:
I confess I didn't read the commit message. 😛 

> I read `-C force-frame-pointers = no` as "Don’t force frame pointers”, rather than anything else

Agreed!

> which aligns well with it simply allowing LLVM to omit the frame pointer anywhere it wants, but not guaranteeing that.

Agreed, but it's not the whole story. I think the part that seems a bit surprising is that not specifying the option *at all* will also 'not force frame pointers', but under different conditions. That is, the behavior of this PR as I understand it:

- `-C force-frame-pointers=yes` -- you get frame pointers
- `-C force-frame-pointers=no` -- you get frame pointers in debug builds, but in opt builds LLVM might optimize frame pointers out
- no option -- you get frame pointers in debug builds, or with debuginfo, or on some platforms, but otherwise LLVM might optimize frame pointers out

I'd be happier if that second and third choice were the same -- I'm not sure which of the two is better. Is there a good reason that they are not the same? e.g., how bad would it be if `-C force-frame-pointers=no` meant the same thing as 'no option' above?